File: LeetCode_Algorithm/217/217.js
var containsDuplicate = function(nums) {
    
  // sol1 n^2 1500~2500ms 11%~4% 보다 빠름
  // for (let i=0; i < nums.length; i++){
  //   for (let j=i+1; j < nums.length; j++){
  //     if (nums[i] == nums[j]){return true}
  //   }
  // }
  // return false

  // sol2 nlogn 280~300ms 20%~16%보다 빠름
  // nums.sort()
  // for (let i=0; i < nums.length-1; i++){
  //   if(nums[i]==nums[i+1]) { return true }
  // }
  // return false

  // sol3 n 84~96ms 95~85%보다 빠름
  const map = {}
  for (let i=0; i < nums.length; i++){
    if ( map[nums[i]]){ return true}
    else{map[nums[i]] = 1}
  }
  return false


};

console.log(containsDuplicate([1,2,3,4,1]))
console.log(containsDuplicate([1,2,3,4]))

File: LeetCode_Algorithm/217(Contains Duplicate)/217.py
from operator import truediv
from typing import List

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:

        # sol1 brute-force n^2 시간초과

        # for i in range(len(nums)):
        #     for j in range(i+1, len(nums)):
        #         if nums[i] == nums[j]: return True
        # return False

        # sol2 nlogn 500~900ms 50% 10%        
        # nlogn으로 sorting한 이후 n으로 읽어가면서 같은 숫자가 보이면 true를 리턴

        # nums.sort()
        # for i in range(len(nums)-1):
        #     if nums[i] == nums[i+1]: return True
        # return False

        # sol3 n 400~500ms 90%~80% 보다 빠름
        # O(n) 해쉬맵을 이용하여 카운트, 해쉬맵에서 존재했다면 true를 리턴
        # checkMap = {}
        # for i in nums:
        #     if checkMap[i]:
        #         checkMap[i] = 1
        #     else:
        #         return True
        # return False
        
        # sol4 n
        # set을 활용하여 set의 길이 == nums의 길이 확인
        return False if len(set(nums))==len(nums) else True



print(Solution().containsDuplicate([1,2,3,4]))